xtm/xtm_api.cc

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "xtm_api.h"
+#include "xtm_scsp_queue.h"
+#include "xtm_config.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <errno.h>
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+#include <sys/eventfd.h>
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+
+#define XTM_PIPE_SIZE 4096
+
+union xtm_msg {
+	/**
+	 * Anonymous structure for storing functions and their arguments,
+	 * when using the xtm API in dispatch/invoke pattern.
+	 */
+	struct {
+		xtm_queue_fun_t fun;
+		void *fun_arg;
+	};
+	/**
+	 * Pointer to save message, when using xtm API
+	 * in send/recv pattern.
+	 */
+	void *ptr;
+};
+
+struct xtm_queue {
+	/**
+	 * File descriptor that the consumer thread must poll,
+	 * to know when new messages are added to the queue.
+	 */
+	int consumer_read_fd;
+	/**
+	 * File descriptor to which the producer thread writes
+	 * to notify the consumer thread of new messages in the queue.
+	 */
+	int consumer_write_fd;
+	/**
+	 * File descriptor that the producer thread must poll,
+	 * to know when there is a free space in qeuue.
+	 */
+	int producer_read_fd;
+	/**
+	 * File descriptor to which the consumer thread writes
+	 * to notify the producer thread about free space in queue.
+	 */
+	int producer_write_fd;
+	/**
+	 * Flag indicates, that producer, wants to get notification
+	 * about free place in queue.
+	 */
+	bool is_producer_need_notifications;
+	/**
+	 * Flag indicates, that we need to close all read file descriptors
+	 * manually, when we delete xtm queue.
```

Comment:
"...that xtm_delete() must close all read end file descriptors, otherwise they must be closed by user before calling xtm_delete()"

"Manually" is misleading here.